registry/storage/driver/swift/swift.go

```diff
@@ -0,0 +1,519 @@
+		Size:    info.Bytes,
+		ModTime: info.LastModified,
+	}
+
+	return storagedriver.FileInfoInternal{FileInfoFields: fi}, nil
+}
+
+// List returns a list of the objects that are direct descendants of the given path.
+func (d *driver) List(ctx context.Context, path string) ([]string, error) {
+	prefix := d.swiftPath(path)
+	if prefix != "" {
+		prefix += "/"
+	}
+
+	opts := &swift.ObjectsOpts{
+		Path:      prefix,
+		Delimiter: '/',
+	}
+
+	files, err := d.Conn.ObjectNames(d.Container, opts)
+	for index, name := range files {
+		files[index] = "/" + strings.TrimSuffix(name, "/")
+	}
+
+	return files, parseError(path, err)
+}
+
+// Move moves an object stored at sourcePath to destPath, removing the original
+// object.
+func (d *driver) Move(ctx context.Context, sourcePath string, destPath string) error {
+	err := d.Conn.ObjectMove(d.Container, d.swiftPath(sourcePath),
+		d.Container, d.swiftPath(destPath))
+	if err != nil {
+		return parseError(sourcePath, err)
+	}
+
+	return nil
+}
+
+// Delete recursively deletes all objects stored at "path" and its subpaths.
+func (d *driver) Delete(ctx context.Context, path string) error {
+	opts := swift.ObjectsOpts{
+		Prefix: d.swiftPath(path),
+	}
+
+	objects, err := d.Conn.ObjectNamesAll(d.Container, &opts)
+	if err != nil {
+		return parseError(path, err)
+	}
+	if len(objects) == 0 {
+		return storagedriver.PathNotFoundError{Path: path}
+	}
+
+	for index, name := range objects {
+		objects[index] = name[len(d.Prefix):]
+	}
+
+	var multiDelete = true
+	if d.BulkDeleteSupport {
+		_, err := d.Conn.BulkDelete(d.Container, objects)
+		multiDelete = err != nil
+	}
+	if multiDelete {
+		for _, name := range objects {
+			if _, headers, err := d.Conn.Object(d.Container, name); err == nil {
+				manifest, ok := headers["X-Object-Manifest"]
+				if ok {
+					components := strings.SplitN(manifest, "/", 2)
+					segContainer := components[0]
+					segments, err := d.Conn.ObjectNamesAll(segContainer, &swift.ObjectsOpts{Prefix: components[1]})
+					if err != nil {
+						return parseError(name, err)
+					}
+
+					for _, s := range segments {
+						if err := d.Conn.ObjectDelete(segContainer, s); err != nil {
+							return parseError(s, err)
+						}
+					}
+				}
+			} else {
+				return parseError(name, err)
+			}
+
+			if err := d.Conn.ObjectDelete(d.Container, name); err != nil {
+				return parseError(name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// URLFor returns a URL which may be used to retrieve the content stored at the given path.
+// May return an UnsupportedMethodErr in certain StorageDriver implementations.
+func (d *driver) URLFor(ctx context.Context, path string, options map[string]interface{}) (string, error) {
+	return "", storagedriver.ErrUnsupportedMethod
+}
+
+func (d *driver) swiftPath(path string) string {
+	return strings.TrimLeft(strings.TrimRight(d.Prefix, "/")+path, "/")
+}
+
+func (d *driver) createParentFolder(path string) (string, error) {
```

Comment:
When PUTting an object, only the first parent directory is created by this function (i.e. the full directory hierarchy is not created)?